cpp/src/parquet/chunker_internal_test.cc

```diff
@@ -0,0 +1,1397 @@
+    AssertChunkSizes(param.dtype, base_info.front(), modified_info.front(),
+                     param.is_nullable, enable_dictionary, kMinChunkSize, kMaxChunkSize);
+    AssertPageLengthDifferences(base_info.front(), modified_info.front(),
+                                /*max_number_of_equal_diffs=*/2);
+  }
+}
+
+TEST_P(TestCDCSingleRowGroup, InsertOnce) {
+  const auto& param = GetParam();
+
+  ASSERT_OK_AND_ASSIGN(auto base, ConcatAndCombine({part1_, part3_}));
+  ASSERT_OK_AND_ASSIGN(auto modified, ConcatAndCombine({part1_, part2_, part3_}));
+  ASSERT_FALSE(base->Equals(*modified));
+
+  for (bool enable_dictionary : {false, true}) {
+    ASSERT_OK_AND_ASSIGN(auto base_info, WriteAndGetParquetInfo(
+                                             base, kMinChunkSize, kMaxChunkSize,
+                                             enable_dictionary, param.data_page_version));
+    ASSERT_OK_AND_ASSIGN(
+        auto modified_info,
+        WriteAndGetParquetInfo(modified, kMinChunkSize, kMaxChunkSize, enable_dictionary,
+                               param.data_page_version));
+    // assert that there is only one row group
+    ASSERT_EQ(base_info.size(), 1);
+    ASSERT_EQ(modified_info.size(), 1);
+    AssertChunkSizes(param.dtype, base_info.front(), modified_info.front(),
+                     param.is_nullable, enable_dictionary, kMinChunkSize, kMaxChunkSize);
+
+    auto edit_length = part2_->num_rows();
+    if (::arrow::is_list_like(param.dtype->id())) {
+      edit_length += 1;
+    }
+    AssertPageLengthDifferences(base_info.front(), modified_info.front(),
+                                /*exact_number_of_equal_diffs=*/0,
+                                /*exact_number_of_larger_diffs=*/1,
+                                /*exact_number_of_smaller_diffs=*/0, edit_length);
+  }
+}
+
+TEST_P(TestCDCSingleRowGroup, InsertTwice) {
+  const auto& param = GetParam();
+
+  ASSERT_OK_AND_ASSIGN(auto base, ConcatAndCombine({part1_, part3_, part5_}));
+  ASSERT_OK_AND_ASSIGN(auto modified,
+                       ConcatAndCombine({part1_, part2_, part3_, part4_, part5_}));
+  ASSERT_FALSE(base->Equals(*modified));
+
+  for (bool enable_dictionary : {false, true}) {
+    ASSERT_OK_AND_ASSIGN(auto base_info, WriteAndGetParquetInfo(
+                                             base, kMinChunkSize, kMaxChunkSize,
+                                             enable_dictionary, param.data_page_version));
+    ASSERT_OK_AND_ASSIGN(
+        auto modified_info,
+        WriteAndGetParquetInfo(modified, kMinChunkSize, kMaxChunkSize, enable_dictionary,
+                               param.data_page_version));
+    // assert that there is only one row group
+    ASSERT_EQ(base_info.size(), 1);
+    ASSERT_EQ(modified_info.size(), 1);
+    AssertChunkSizes(param.dtype, base_info.front(), modified_info.front(),
+                     param.is_nullable, enable_dictionary, kMinChunkSize, kMaxChunkSize);
+
+    auto edit_length = part2_->num_rows();
```

Comment:
We're also inserting `part4_`, it doesn't matter? Or is it the same length? These tests seem to be relying on hidden assumptions, which will make them difficult to maintain.

Reply:
Added a comment about this in the assertion function.